don't use PSFs with bad amps

sbailey opened the pull request:
This PR fixes #1471 by updating desi_proc to not use PSFs that have a masked bad amp, e.g. recent r8A.

It still generates the PSF file for post-facto debugging, but if the input CCD has a flagged amp it renames the psf to (origname).badreadnoise and skips the bad fiber interpolation step.  Because this doesn't result in a "fit-psf*.fits" file, it won't get picked up by the merging step.

Implementation detail: although currently we flag entire amps with the ccdmask.BADREADNOISE bit, I implemented this as a somewhat more generic >12.5% pixels (half of one amp) have BADREADNOISE set, just in case some future update changes how BADREADNOISE is set (I realize that readnoise physically comes from individual amps, but sometimes it "turns on" or off during readout and we also have effective readnoise coming from the bias etc...)

Also coming along for the ride was a fix to `proc_joint_fit.main` which previously was doing an over-enthusiastic string replacement that would break any $SPECPROD containing "psf".

I tested this in /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/badpsf with night 20211026 which ended up rejecting all r8 PSFs because all 5 input arcs had r8 OSTEPA > 5.  Jobs are in the queue to test 20211025 (which should reject one arc but accept the other four and proceed with a nightly PSF) and 20211024 (which should accept all five arcs).

sbailey commented:
Tests finished on 20211024, 1025, and 1026 in /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/badpsf with results as expected for r8:
* 20211024: no arcs rejected, psfnight-r8-20211024.fits generated
* 20211025: 1/5 arcs rejected, psfnight-r8-20211025.fits generated with remaining 4
* 20211026: 5/5 arcs rejected, psfnight-r8-20211026.fits *not* generated (but other cameras have psfnight)

Minor change after chatting with Julien: changed threshold to rejecting a CCD if half of one amp (0.5 * 0.25 = 12.5%) of pixels are flagged BADREADNOISE, instead of 5%.  Makes no change for current code where we flag entire amps as bad, but provides some future proofing if that changes.

Ready for review.

akremin commented:
Also noting this throws out an entire camera due to one bad amplifier. Previously @julienguy had been advocating for just throwing out the problematic fibers, or not throwing them out at all for the cals since the counts are substantially higher than these jumps?

The logic used in assemble fibermap for a badamp is here, if useful:
https://github.com/desihub/desispec/blob/e0068dd8603444ab21c994d29ab15b8d35cbf8a2/py/desispec/io/fibermap.py#L778

akremin commented:
I'm confused how night 20211025 produced a psfnight given that the code currently requires more than 4 psfs for a given camera to run: https://github.com/desihub/desispec/blob/e0068dd8603444ab21c994d29ab15b8d35cbf8a2/py/desispec/scripts/proc_joint_fit.py#L185

I think this can be changed in the code to run for e.g. psfs >3. The comment is more in surprise to the result of the test

sbailey commented:
psf vs. fit-psf vs. findfile('fitpsf', ...) -- this is a bit of a mess in the code pre-dating this PR.  Standardizing everything via findfile would be better, but I'm a little hesitant to do that just for fitpsf without also diving into "fit-psf-before-listed-fix" and "fit-psf-fixed-listed" and cleaning up writing files by one name and then renaming in certain cases and copying in others...  Thinking... e.g. in particular I don't want to end up in a situation where one line is using `findfile('fitpsf', ...)` while another line is assuming that returns a `fit-psf` prefix for the purposes of some other replacement to `fit-psf-fixed-listed`...

Regarding `if len(psfs) > 4: ...` -- that code in proc_joint_fit is counting how many PSFs could exist based upon expids without actually checking if they do exists, leaving it to merge_psf to decide what to do with missing files.  I agree with having some minimum, but will come chat about whether proc_joint_fit should be enforcing that or whether merge_psf should be enforcing that...

sbailey commented:
I updated desi_proc_joint_fit to be the gatekeeper of how many PSFs it wants before passing to merge PSF.  Output now is:
```
...
TIMER-START:proc_joint_fit.py:177:main: Starting psfnight at 2021-11-03T16:36:30.477441
WARNING:proc_joint_fit.py:189:main: Missing /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/badpsf/exposures/20211025/00105975/fit-psf-r8-00105975.fits
INFO:proc_joint_fit.py:191:main: Number of PSF for night=20211025 camera=r8 = 4/5
INFO:proc_joint_fit.py:194:main: Computing r8 psfnight ...
INFO:specex.py:434:mean_psf: Will compute the average of 4 PSFs
...
```

I switched proc_joint_fit to use `findfile('fitpsf', ...)` but postponed any further cleanup of the chain of flavors of psf fit files in desi_proc itself.
